backend/compact-connect/stacks/persistent_stack/ssn_table.py

```diff
@@ -0,0 +1,157 @@
+    Role,
+    ServicePrincipal,
+    StarPrincipal,
+)
+from aws_cdk.aws_kms import Key
+from cdk_nag import NagSuppressions
+from common_constructs.stack import Stack
+from constructs import Construct
+
+
+class SSNTable(Table):
+    """DynamoDB table to house provider Social Security Numbers"""
+
+    def __init__(
+        self,
+        scope: Construct,
+        construct_id: str,
+        *,
+        removal_policy: RemovalPolicy,
+        **kwargs,
+    ):
+        self.key = Key(
+            scope,
+            'SSNKey',
+            enable_key_rotation=True,
+            alias='ssn-key',
+            removal_policy=removal_policy,
+        )
+
+        super().__init__(
+            scope,
+            construct_id,
+            # Forcing this resource name to comply with a naming-pattern-based CloudTrail log, to be
+            # implemented in issue https://github.com/csg-org/CompactConnect/issues/397
+            table_name='ssn-table-DataEventsLog',
+            encryption=TableEncryption.CUSTOMER_MANAGED,
+            encryption_key=self.key,
+            billing_mode=BillingMode.PAY_PER_REQUEST,
+            removal_policy=removal_policy,
+            point_in_time_recovery=True,
+            partition_key=Attribute(name='pk', type=AttributeType.STRING),
+            sort_key=Attribute(name='sk', type=AttributeType.STRING),
+            resource_policy=PolicyDocument(
+                statements=[
+                    PolicyStatement(
+                        # No actions that involve reading/writing more than one record at a time. In the event of a
+                        # compromise, this slows down a potential data extraction, since each record would need to be
+                        # pulled, one at a time
+                        effect=Effect.DENY,
+                        actions=[
+                            'dynamodb:BatchGetItem',
+                            'dynamodb:BatchWriteItem',
+                            'dynamodb:PartiQL*',
+                            'dynamodb:Query',
+                            'dynamodb:Scan',
+                        ],
+                        principals=[StarPrincipal()],
+                        resources=['*'],
+                        conditions={
+                            'StringNotEquals': {
+                                # We will allow DynamoDB itself, so it can do internal operations like backups
+                                'aws:PrincipalServiceName': 'dynamodb.amazonaws.com',
+                            }
+                        },
+                    )
+                ]
+            ),
+            **kwargs,
+        )
+
+        # We create a GSI here in anticipation of a future change, where we will need to facilitate a lookup
+        # of provider_id -> ssn, in addition to our current ssn -> provider_id pattern.
+        self.inverted_index_name = 'inverted'
+        self.add_global_secondary_index(
+            index_name=self.inverted_index_name,
```

Comment:
Dumb question...but where does the 'inverted' part of the key show up in the codebase?

Reply:
It doesn't yet. This index is one we will end up using in the future. I wanted to create it ahead of time so we didn't have to add an index to an existing table later on.

Reply:
Or sorry-- I meant to say, "Where, if anywhere, is the _value_ of the GSI set? i.e., where it sets which field(s) are indexed."  

Maybe the answer is the same, still?

Reply:
That's what this function call is doing - declaring a GSI and defining its partition key, sort key, and projected attributes.

Reply:
Ah, I missed the flipped sk/pk fields. Small nit, but can we just add a brief comment saying that to facilitate the reverse lookup, we reverse the primary and sort keys?